libpromises/promises.c

```diff
@@ -240,6 +250,19 @@ Promise *DeRefCopyPromise(EvalContext *ctx, const Promise *pp)
 
                         if (IsDefinedClass(ctx, scp->classes))
                         {
+                            SeqAppend(constraints_to_expand, scp);
+                        }
+                    }
+
+                    SeqReverse(constraints_to_expand);
+                    // Only do the copy expansion if we copied something while inheriting
+                    for (size_t r = 0; r < SeqLength(constraints_to_expand); r++)
```

Comment:
The logic is a bit difficult to follow, is the reason for reversing that you need to resolve the constraints with the "youngest" body taking precedence?

In any case, this will have better performance if you don't reverse the sequence (especially since you reverse it back later), but rather reverse the loop, starting at `SeqLength(constraints_to_expand) - 1` and going to `0`.